fix: Infinity scroll is slow

Jarsen136 opened the pull request:
**Thank you for your contribution** to the [KodaDot NFT gallery](https://kodadot.xyz).

👇 \_ Let's make a quick check before the contribution.

## PR Type

- [x] Bugfix
- [ ] Feature
- [ ] Refactoring

## Context

- [x] Closes #5394
- [x] Closes #5461 
- [x] Closes #5457
- [ ] Requires deployment <>

#### Before submitting pull request, please make sure:

- [x] My contribution builds **clean without any errors or warnings**
- [x] I've merged recent default branch -- **main** and I've no conflicts
- [x] I've tried to respect high code quality standards
- [x] I've didn't break any original functionality
- [ ] I've posted a screenshot of demonstrated change in this PR

#### Optional

- [ ] I've tested it at </bsx/collection>
- [ ] I've tested PR on mobile
- [ ] I've written unit tests 🧪
- [ ] I've found edge cases

#### Had issue bounty label?

- [x] Fill up your KSM address: [Payout](https://beta.kodadot.xyz/transfer/?target=Caiv9TbPz68q5dC8EcHu5xKYPRnremimGzqmEejDFNpWWLG)

#### Community participation

- [x] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

## Screenshot 📸
tested on `/rmrk/explore/items?page=6`

roiLeo reviewed (commented):
Would it not be better to integrate [nuxt-img](https://image.nuxtjs.org/components/nuxt-img) instead of `vue-lazyload`?

I ask this because I would like to avoid adding new plugins until we've done Nuxt3 migration

Jarsen136 commented:
> > Would it not be better to integrate [nuxt-img](https://image.nuxtjs.org/components/nuxt-img) instead of `vue-lazyload`?
> > I ask this because I would like to avoid adding new plugins until we've done Nuxt3 migration
> 
> Thanks. I will take a look.

I have tried to integrate nuxt-image into our project. It does works, but I found there still be something strange left to handler.

-  ipx provider https://image.nuxtjs.org/providers/ipx
-  full image url -> relative image url
- `nuxt.config.js`  the `target` config should be stay `static` instead of `server`
- related error:  
<img width="726" alt="image" src="https://user-images.githubusercontent.com/31397967/229035584-85b1f511-0c0d-4e06-96c7-ebdf6f69dfc6.png">


 I guess we could make another issue for someone else to pick it. Here is the demo PR
https://github.com/kodadot/nft-gallery/pull/5440

roiLeo commented:
> I have tried to integrate nuxt-image into our project. It does works, but I found there still be something strange left to handler.

thank you for your investigation and sorry to see that it did not worked

> ipx provider https://image.nuxtjs.org/providers/ipx

Is this the solution for our case? Maybe custom provider is more in line with our needs?

> full image url -> relative image url

from what i understand images are served by the same domain `https://image(-beta).w.kodadot.xyz/...`

> `nuxt.config.js`  the `target` config should be stay `static` instead of `server`

yep this changes implies many modifications and could break stuff

> I guess we could make another issue for someone else to pick it. Here is the demo PR #5440

maybe #4426?


btw in #5440 I can see that you use nuxt-image in `libs/ui` or this folder shouldn't have access to nuxt (not installed over here)

roiLeo reviewed (commented):
so you're using `v-lazy` in `libs/ui`? How would it worked if we remove this package from kodadot repo?

vikiival commented:
Agree w @preschian 

@Jarsen136 please remove vue-lazy, we will make it in separated PR

Jarsen136 commented:
> > I have tried to integrate nuxt-image into our project. It does works, but I found there still be something strange left to handler.
> 
> thank you for your investigation and sorry to see that it did not worked
> 
> > ipx provider https://image.nuxtjs.org/providers/ipx
> 
> Is this the solution for our case? Maybe custom provider is more in line with our needs?
> 
Yes, it's the solution for the incorrect URL prefix.

> > full image url -> relative image url
> 
> from what i understand images are served by the same domain `https://image(-beta).w.kodadot.xyz/...`
> 
Yup, but when I set `static` to the target in `nuxt.config.js` file, the nuxt-image would not work with full image url like `https://image(-beta).w.kodadot.xyz/{id}`. It need the src to at the format like `/privoide/{id}`.

> > `nuxt.config.js`  the `target` config should be stay `static` instead of `server`
> 
> yep this changes implies many modifications and could break stuff
> 

> > I guess we could make another issue for someone else to pick it. Here is the demo PR #5440
> 
> maybe #4426?
> 
> btw in #5440 I can see that you use nuxt-image in `libs/ui` or this folder shouldn't have access to nuxt (not installed over here)

There are more details that could be continued by #4426

Jarsen136 commented:
> > Agree w @preschian
> > @Jarsen136 please remove vue-lazy, we will make it in separated PR
> 
> OK, I will make it.

Updated. I have removed the vue-lazyload plugin. Though scrolling process would be more smooth with the help of image lazyload.

yangwao commented:
seems it's doing weird gltiches when scrolling down and returning me back? 

https://user-images.githubusercontent.com/5887929/229222786-fc4963d1-8a7f-4c12-b5c3-91d95fda4951.mov

Jarsen136 commented:
> seems it's doing weird gltiches when scrolling down and returning me back?

I could not reproduce that bug may because it would happen on a higher screen. And I make a fix based on that strange behaviour. Is it better now?

Jarsen136 commented:
Should fix all of these issues #5394 #5457 #5461.

Jarsen136 commented on libs/ui/src/components/MediaItem/type/ImageMedia.vue:
```diff
@@ -6,10 +6,12 @@
     <img
-      class="is-block image-media__image"
       :src="src"
+      loading="lazy"
```
ok

vikiival commented:
<img width="682" alt="Screenshot 2023-04-02 at 22 28 02" src="https://user-images.githubusercontent.com/22471030/229377365-d353a90a-cc04-44f7-a4df-9afbd1201374.png">


👀 You just used your one-per-month-force-push token

vikiival commented on libs/ui/src/components/MediaItem/type/ImageMedia.vue:
```diff
@@ -20,6 +21,19 @@ defineProps<{
+const onError = (e: Event) => {
+  const target = e.target as HTMLImageElement
+  if (target) {
+    $consola.log('image load error', e)
```
```suggestion
    $consola.log('[KODADOT::IMAGE] unable to load', e)
```

vikiival commented on composables/useListInfiniteScroll.ts:
```diff
@@ -44,11 +44,11 @@ export default function ({
       }
     },
-    { distance: 1600 }
+    { distance: 0 }
```
```
   * The minimum distance between the bottom of the element and the bottom of the viewport
   *
   * @default 0
   */
  distance?: number
  ```

Jarsen136 replied:
Thanks, I have remove it.

vikiival commented on composables/useListInfiniteScroll.ts:
```diff
@@ -35,7 +35,7 @@ export default function ({
   const itemsPerRow = ref(4)
   const scrollItemSizeInit = ref(false)
-  const first = ref(defaultFirst || 20)
+  const first = ref(defaultFirst || 40)
```
Isn't this causing it too slow? 

Loading 40 items takes quuite a memory

Jarsen136 replied:
I have discovered that the time required to fetch 20 items versus fetching 40 items is quite similar. Both of them take 500ms or so. Or I would limit it to 30. WDYT

vikiival replied:
Ok if it's like that I am fine. 
Or let's do 30 and we will see

Jarsen136 commented:
> <img alt="Screenshot 2023-04-02 at 22 28 02" width="682" src="https://user-images.githubusercontent.com/22471030/229377365-d353a90a-cc04-44f7-a4df-9afbd1201374.png">
> 
> 👀 You just used your one-per-month-force-push token

My bad. It's because I pushed something unrelated commit to this branch by mistake. : )

roiLeo commented on libs/ui/src/components/MediaItem/type/ImageMedia.vue:
```diff
@@ -20,6 +21,19 @@ defineProps<{
   original: boolean
   isDetail?: boolean
 }>()
+const { $consola } = useNuxtApp()
+const { isDarkMode } = useTheme()
```
This shouldn't be used here

Jarsen136 replied:
Yes, I have remove these stuff

roiLeo commented on libs/ui/src/components/MediaItem/type/ImageMedia.vue:
```diff
@@ -20,6 +21,19 @@ defineProps<{
 }>()
+const { $consola } = useNuxtApp()
+const { isDarkMode } = useTheme()
+const placeholder = computed(() => {
+  return isDarkMode.value ? '/placeholder.webp' : '/placeholder-white.webp'
+})
+
+const onError = (e: Event) => {
+  const target = e.target as HTMLImageElement
+  if (target) {
+    $consola.log('[KODADOT::IMAGE] unable to load', e)
+    target.src = placeholder.value
+  }
+}
```
I would move this at component level

Jarsen136 replied:
The error is throw directly from the img tag, so I guess placing an error handler on this component is not bad.

exezbcz commented:
hmm, cannot open the deploy link, Am i doing something wrong? 
![image](https://user-images.githubusercontent.com/90852205/229513721-aa849f1a-ac95-452e-9b5b-f9681b56b39b.png)

vikiival commented:
@yangwao 👀

yangwao commented:
> hmm, cannot open the deploy link, Am i doing something wrong?

https://deploy-preview-5429--koda-canary.netlify.app/ 

been changing deployments and old PRs might get wacky, `koda-nuxt` changed to `koda-canary`, at least know netlify retroactive doesn't update :p 

- #5420 

changing deployment while things are running 
![image](https://user-images.githubusercontent.com/5887929/229521667-e5f61fb5-db4b-4499-9efa-4c5a04db446b.png)

vikiival commented:
Lets fix @roiLeo stuff and let it roll

yangwao commented:
wow amazing fix @Jarsen136 seems infinity scrolling is back to great state! :)
thank you!

Jarsen136 commented:
> Lets fix @roiLeo stuff and let it roll

✅  Updated

roiLeo reviewed (commented):
why do you need to access `darkMode` in `libs/ui`?
same for `placeholder` I would pass it as props instead.

Use case: 
I'm using `@kodadot1/brick` in a different project and I want to use custom placeholder

yangwao commented:
can we merge it 🥹

vikiival commented:
Yup let it roll, and @roiLeo issue will be resolved in next one
